src/utils/factory.hpp

```diff
@@ -0,0 +1,91 @@
+/*
+ * Authors: M. Kristien
+ * Organisation: HYPED
+ * Date: Dec 2019
+ * Description:
+ *
+ *    Copyright 2019 HYPED
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#ifndef UTILS_FACTORY_HPP_
+#define UTILS_FACTORY_HPP_
+/**
+ * @brief An abstract templated factory to provide mapping from string names
+ * of abstract interfaces to functions creating the implementations of the interfaces.
+ *
+ * The factory is templated with the interface class. This means the factory is created
+ * individually for different interfaces. Different implementations of the same interface
+ * use the same factory.
+ */
+
+#include <map>
+#include <string>
+
+#include "utils/interfaces.hpp"
+#include "utils/logger.hpp"
+
+using std::string;
+
+namespace hyped {
+namespace utils {
+
+template<class Interface>
+class Factory {
+  typedef Interface* (*InterfaceCreator)();
```

Comment:
What does this mean/do? I genuinely have no idea what I'm looking at :sweat_smile:

Reply:
This is a function pointer. Basically a variable to which you can assign function. I name the type of the variable as `InterfaceCreator` here. The type can only be assigned functions with signature like: `Interface* <name>()`.